### PR TITLE
[ARM32/Linux] Build libcoreclrtraceptprovider.so

### PR DESCRIPTION
**Pull request by hseok-oh**
- Build libcoreclrtraceptprovider.so for ARM32
- Include libcoreclrtraceptprovider.so in nuget package
- Except Tizen(armel): lttng is not supported in Tizen

related issue: #8684

### CODE REVIEW

**Comment by hseok-oh** [comment]
@dotnet-bot test Ubuntu arm Cross Release Build 
@dotnet-bot test Ubuntu16.04 arm Cross Debug Build
@dotnet-bot test Windows_NT x86 Checked Build and Test

File: CMakeLists.txt [hunk 1]
```diff
@@ -161,14 +161,25 @@ if(CMAKE_SYSTEM_NAME STREQUAL Linux)
     endif()
     set(CLR_CMAKE_PLATFORM_LINUX 1)
 
-    # Detect Alpine Linux
-    SET(OS_RELEASE_FILENAME "/etc/os-release")
+    # Detect Alpine & Tizen Linux
+    if(DEFINED ENV{CROSSCOMPILE})
```
Comment (janvorli) [inline CMakeLists.txt]: I prefer not to use env vars as switches controlling cmake scripts. Especially here where we alredy have a similar switch passed by CLR_CROSS_COMPONENTS_BUILD cmake variable. 
Could you please change it to use CLR_CROSSCOMPILE cmake variable and set / pass it to cmake in `gen-buildsys-clang.sh` similarly to how we set / pass the CLR_CROSS_COMPONENTS_BUILD?

File: src/.nuget/Microsoft.NETCore.Runtime.CoreCLR/runtime.Linux.Microsoft.NETCore.Runtime.CoreCLR.props [hunk 2]
```diff
@@ -6,7 +6,7 @@
   </PropertyGroup>
   <ItemGroup>
     <NativeBinary Include="$(BinDir)libcoreclr.so" />
-    <NativeBinary Condition="'$(_PlatformDoesNotSupportNiFiles)' != 'true'" Include="$(BinDir)libcoreclrtraceptprovider.so" />
+    <NativeBinary Condition="'$(Platform)' != 'armel'" Include="$(BinDir)libcoreclrtraceptprovider.so" />
```
Comment (janvorli) [inline src/.nuget/Microsoft.NETCore.Runtime.CoreCLR/runtime.Linux.Microsoft.NETCore.Runtime.CoreCLR.props]: It looks strange to conditionalize it here based on the platform when you really mean Tizen. People can do armel builds for other platforms where the trace provider can be built, right?
Actually, the previous condition was also strange.
@gkhanna79, can you see a better way to conditionalize the inclusion of the libcoreclrtraceptprovider.so in the package?

**Comment by hseok-oh** [comment]
I've reflected @janvorli 's review. PTAL.

- Path to cmake cross-compilation and linux id using argument
CLR_CROSS_COMPILE=1 when cross-compile
CLR_CMAKE_LINUX_ID as linux id from $rootfs_dir/etc/os-release

- Exclude packaging libcoreclrtraceptprovider.so based on os-name
excluded for tizen only

**Comment by hseok-oh** [comment]
@janvorli Thanks for fast review. :)

**Comment by hseok-oh** [comment]
@dotnet-bot test Ubuntu arm Cross Release Build please

**Comment by janvorli** [comment]
@dotnet-bot test Ubuntu arm Cross Release Build please

**Comment by hseok-oh** [comment]
@janvorli It seems that we need to change docker image for CoreFX CI build first to avoid test failure. If this is merged first, we may suffers CI test failure.
I've made new PR (dotnet/corefx#17541) for this and I'll ping you if that PR merged.

**Comment by hseok-oh** [comment]
@dotnet-bot test Ubuntu arm Cross Release Build please

**Comment by hseok-oh** [comment]
@dotnet-bot test Ubuntu arm Cross Release Build please

**Comment by hseok-oh** [comment]
@dotnet-bot test Ubuntu arm Cross Debug Build

**Comment by hseok-oh** [comment]
@dotnet-bot test Ubuntu arm Cross Debug Build
@dotnet-bot test Ubuntu16.04 arm Cross Release Build

**Comment by hseok-oh** [comment]
@dotnet-bot test Windows_NT x64 Release Priority 1 Build and Test please

**Comment by hseok-oh** [comment]
@janvorli I've modified some points. PTAL.

- Don't build libcoreclrtraceptprovider.so for ubuntu 14.04 arm
- CMake argument set CLR_CMAKE_LINUX_VERSION as os version from /etc/os-release
- Exclude packaging libcoreclrtraceptprovider.so for ubuntu 14.04 arm

We can build libcoreclrtraceptprovider.so for ubuntu 14.04 arm, but it seems to have problem: test failure. But when I tested by removing this file in my environment (using docker and qemu for test as CI does), the failure cleared. And when I tested in ubuntu 16.04 arm device (raspberry pi 3) using coreclr build for ubuntu 14.04 arm with libcoreclrtraceptprovider.so, the failure also cleared. Build for ubuntu 16.04 arm has no problem. 

Unfortunately I don't have any test device to test ubuntu 14.04 arm. It seems to need more investigation. And it's better to remain new issue.

**Comment by BruceForstall** [comment]
@hseok-oh Looks like you are still working on this, due to test failures? Do you still need @janvorli review?

**Comment by hseok-oh** [comment]
@BruceForstall I'm still working on this and I suppose it's docker + qemu emulation problem. I'll request review again if I check the results in ubuntu 14.04 arm device soon.

**Comment by hseok-oh** [comment]
@dotnet-bot test Tizen armel Cross Release Build

**Comment by hseok-oh** [comment]
Our test environment uses docker and qemu-static. We don't use qemu-bootstrap, instead we use mounting manually some system directory and pre-generated rootfs for performance. But it seems to have problem to use library lttng.
I confirmed this commit work well in raspberry pi 2 (ubuntu 14.04 armhf). So I generate libcoreclrtraceptprovider.so and remove for test.

@janvorli PTAL

**Comment by wtgodbe** [comment]
@hseok-oh @janvorli I believe this change is causing failures in the pipeline during Ubuntu14.04 Crossbuilds: https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_build/explorer?_a=summary&buildId=675849

**Comment by wtgodbe** [comment]
> 2017-04-13T15:26:07.6933220Z 4 errors generated.
2017-04-13T15:26:07.6965770Z make[2]: *** [src/pal/src/eventprovider/tracepointprovider/CMakeFiles/coreclrtraceptprovider.dir/__/lttng/traceptprovdotnetruntimerundown.cpp.o] Error 1
2017-04-13T15:26:07.6979480Z make[2]: *** Waiting for unfinished jobs....
2017-04-13T15:26:07.7360750Z In file included from /root/coreclr/bin/obj/Linux.arm.Checked/Generated/eventprovider/lttng/traceptprovdotnetruntimestress.cpp:15:
2017-04-13T15:26:07.7374670Z In file included from /root/coreclr/bin/obj/Linux.arm.Checked/Generated/eventprovider/lttng/./tpdotnetruntimestress.h:129:
2017-04-13T15:26:07.7389300Z In file included from /crossrootfs/arm/usr/include/lttng/tracepoint-event.h:58:
2017-04-13T15:26:07.7404610Z In file included from /crossrootfs/arm/usr/include/lttng/ust-tracepoint-event.h:26:
2017-04-13T15:26:07.7422160Z In file included from /crossrootfs/arm/usr/include/urcu/rculist.h:30:
2017-04-13T15:26:07.7441110Z In file included from /crossrootfs/arm/usr/include/urcu-pointer.h:31:
2017-04-13T15:26:07.7455340Z In file included from /crossrootfs/arm/usr/include/urcu/uatomic.h:41:
2017-04-13T15:26:07.7470480Z /crossrootfs/arm/usr/include/urcu/uatomic/generic.h:75:10: error: address argument to atomic builtin must be a pointer to integer or pointer ('void *' invalid)
2017-04-13T15:26:07.7485120Z                 return __sync_val_compare_and_swap_4(addr, old, _new);
2017-04-13T15:26:07.7498150Z                        ^                             ~~~~
2017-04-13T15:26:07.7553840Z /crossrootfs/arm/usr/include/urcu/uatomic/generic.h:112:3: error: address argument to atomic builtin must be a pointer to integer or pointer ('void *' invalid)
2017-04-13T15:26:07.7568560Z                 __sync_and_and_fetch_4(addr, val);
2017-04-13T15:26:07.7581520Z                 ^                      ~~~~
2017-04-13T15:26:07.7594670Z Scanning dependencies of target dlltest2_exports
2017-04-13T15:26:07.7611550Z /crossrootfs/arm/usr/include/urcu/uatomic/generic.h:151:3: error: address argument to atomic builtin must be a pointer to integer or pointer ('void *' invalid)
2017-04-13T15:26:07.7663950Z                 __sync_or_and_fetch_4(addr, val);
2017-04-13T15:26:07.7677330Z                 ^                     ~~~~
2017-04-13T15:26:07.7692070Z /crossrootfs/arm/usr/include/urcu/uatomic/generic.h:190:10: error: address argument to atomic builtin must be a pointer to integer or pointer ('void *' invalid)
2017-04-13T15:26:07.7706930Z                 return __sync_add_and_fetch_4(addr, val);
2017-04-13T15:26:07.7720080Z                        ^                      ~~~~
2017-04-13T15:26:07.7915760Z [ 18%] [ 18%] Building CXX object src/vm/dac/CMakeFiles/cee_dac.dir/__/peimage.cpp.o
2017-04-13T15:26:07.8037670Z 4 errors generated.
2017-04-13T15:26:07.8057780Z make[2]: *** [src/pal/src/eventprovider/tracepointprovider/CMakeFiles/coreclrtraceptprovider.dir/__/lttng/traceptprovdotnetruntimestress.cpp.o] Error 1

**Comment by gkhanna79** [comment]
This resulted in break because CoreCLR is using V2 Docker image for 1404 in official build. It should use V3, similar to what CoreFX was updated to do - see https://github.com/dotnet/corefx/pull/17541/files.

@wtgodbe Can you please make the change and queue off another build?

**Comment by wtgodbe** [comment]
It has been done - https://devdiv.visualstudio.com/DevDiv/Net%20CoreFX/_build/index?buildId=676675&_a=summary

https://github.com/dotnet/coreclr/pull/10949

**Comment by gkhanna79** [comment]
Thanks @wtgodbe - please do keep an eye on the build.

**Comment by wtgodbe** [comment]
Looks like that worked! All the 14.04 builds succeeded - https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_build/explorer?_a=completed&definitionId=5972

**Comment by wtgodbe** [comment]
@hseok-oh @janvorli you can disregard my earlier comment

**Pull request merged**